### PR TITLE
Simplifying Steel effect hierarchy

### PR DESCRIPTION
**Pull request by R1kM**
This PR simplifies the Steel effect hierarchy by removing the need for a 2-effect system and polymonads to encode framing.
Instead, Steel/SteelF have a unique representation, with a semantically irrelevant boolean which indicates whether the computation has already been framed. In this case, the frame is resolved to emp at the very beginning of the tactic, using the additional `maybe_emp(_dep)` constraint.

This change is also applied to SteelAtomic and SteelSel, and relations between SteelAtomic and Steel is simplified: instead of polymonadic subcomps and a multitude of polymonadic binds, we simply have a lift from SteelAtomic to Steel.

For now, composing pure computations with Steel* computations still requires polymonadic binds. An attempt at replacing this by a lift is commented out in Steel.Effect.fsti, but it does not allow dependent returns. We would need a post-parametric lift for this, but we cannot bind the value returned by a pure computation `f` in the repr indices when lifting `f`.

### CODE REVIEW

**Comment by R1kM** [comment]
After yesterday's Steel meeting, it was decided that the PURE, Steel* polymonadic binds were ok to keep

File: ulib/experimental/Steel.Channel.Simplex.fst [hunk 1]
```diff
@@ -161,7 +161,7 @@ let intro_in_state (r:ref chan_val) (p:prot) (v:chan_val_p p)
   = intro_pure (in_state_prop p v);
     intro_exists v (fun (v:chan_val) -> pts_to r half v `star` in_state_slprop p v)
 
-#push-options "--print_universes"
+//#push-options "--print_universes"
```
Comment (aseemr) [inline ulib/experimental/Steel.Channel.Simplex.fst:164]: Delete?

File: ulib/experimental/Steel.Channel.Simplex.fst [hunk 2]
```diff
@@ -273,8 +273,9 @@ let gather_r (#p:sprot) (r:ref chan_val) (v:chan_val)
     (fun _ -> pts_to r full_perm v `star` in_state_slprop p v)
   = let v' = witness_h_exists () in
     Steel.Utils.higher_ref_pts_to_injective_eq r v _;
-    H.gather r;
-    rewrite_context()
+    H.gather #_ #half #half #v #v r;
+    change_slprop (pts_to r (sum_perm half half) v) (pts_to r full_perm v) (fun _ -> ());
+    change_slprop (in_state_slprop p v') (in_state_slprop p v) (fun _ -> ())
```
Comment (aseemr) [inline ulib/experimental/Steel.Channel.Simplex.fst:278]: Any ideas why these changes? Are the constraints different and we regressed a bit in solving them?

File: ulib/experimental/Steel.Effect.Atomic.fsti [hunk 3]
```diff
@@ -155,25 +167,27 @@ let if_then_else_ens (#a:Type) (#pre_f:pre_t) (#pre_g:pre_t) (#post_f:post_t a)
...
   (#[@@@ framing_implicit] pre_f:pre_t) (#[@@@ framing_implicit] pre_g:pre_t)
   (#[@@@ framing_implicit] post_f:post_t a) (#[@@@ framing_implicit] post_g:post_t a)
   (#[@@@ framing_implicit] req_then:req_t pre_f) (#[@@@ framing_implicit] ens_then:ens_t pre_f a post_f)
   (#[@@@ framing_implicit] req_else:req_t pre_g) (#[@@@ framing_implicit] ens_else:ens_t pre_g a post_g)
   (#[@@@ framing_implicit] s_pre: squash (can_be_split pre_f pre_g))
   (#[@@@ framing_implicit] s_post: squash (equiv_forall post_f post_g))
-  (f:repr a o unobservable pre_f post_f req_then ens_then)
-  (g:repr a o unobservable pre_g post_g req_else ens_else)
+  (f:repr a framed o unobservable pre_f post_f req_then ens_then)
+  (g:repr a framed o unobservable pre_g post_g req_else ens_else)
```
Comment (aseemr) [inline ulib/experimental/Steel.Effect.Atomic.fsti:178]: Can you please remind me (and also perhaps add a comment?) why do we require both to have the same framed bit? E.g. why can't we say that both may have different framed bits, and then the final computation is not framed?

File: ulib/experimental/Steel.Effect.Atomic.fsti [hunk 4]
```diff
@@ -375,224 +255,38 @@ val bind_pure_steela_ (a:Type) (b:Type)
...
     pre
     post
     (bind_pure_steela__req wp req)
     (bind_pure_steela__ens wp ens)
 
-polymonadic_bind (PURE, SteelAtomicF) |> SteelAtomicF = bind_pure_steela_
+polymonadic_bind (PURE, SteelAtomicBase) |> SteelAtomicBase = bind_pure_steela_
 
-polymonadic_bind (PURE, SteelAtomic) |> SteelAtomic = bind_pure_steela_
+// polymonadic_bind (PURE, SteelAtomic) |> SteelAtomic = bind_pure_steela_
```
Comment (aseemr) [inline ulib/experimental/Steel.Effect.Atomic.fsti:271]: Delete? (and below as well, line 273.)

File: ulib/experimental/Steel.Effect.Common.fsti [hunk 5]
```diff
@@ -1608,8 +1639,8 @@ let init_resolve_tac () : Tac unit =
   // To debug, it is best to look at the goals at this stage. Uncomment the next line
   // dump "initial goals";
 
-  // We now solve all postconditions of pure returns to avoid restricting the uvars
-  solve_all_returns (goals ());
+  // // We now solve all postconditions of pure returns to avoid restricting the uvars
+  // solve_all_returns (goals ());
```
Comment (aseemr) [inline ulib/experimental/Steel.Effect.Common.fsti:1643]: Delete this too? Also curious how come we don't need it anymore?

File: ulib/experimental/Steel.Effect.fsti [hunk 6]
```diff
@@ -374,33 +209,48 @@ let bind_pure_steel__ens (#a:Type) (#b:Type)
...
     pre
     post
     (bind_pure_steel__req wp req)
     (bind_pure_steel__ens wp ens)
 
-polymonadic_bind (PURE, SteelF) |> SteelF = bind_pure_steel_
+polymonadic_bind (PURE, SteelBase) |> SteelBase = bind_pure_steel_
 
-polymonadic_bind (PURE, Steel) |> Steel = bind_pure_steel_
+// AF: The lift below works, but the framing tactic currently requires annotations
```
Comment (aseemr) [inline ulib/experimental/Steel.Effect.fsti:225]: May be clean it up from here and add a separate note somewhere in wiki or github or fstar-priv?

File: ulib/experimental/Steel.Effect.fsti [hunk 7]
```diff
@@ -374,33 +209,48 @@ let bind_pure_steel__ens (#a:Type) (#b:Type)
...
 
+// sub_effect PURE ~> SteelBase = lift_pure_steel
 
 (*
 //  * Annotations without the req and ens
 //  *)
 
 effect SteelT (a:Type) (pre:pre_t) (post:post_t a) =
-  Steel a pre post (fun _ -> True) (fun _ _ _ -> True)
+  SteelBase a false pre post (fun _ -> True) (fun _ _ _ -> True)
```
Comment (aseemr) [inline ulib/experimental/Steel.Effect.fsti:253]: For readability you could define it as `Steel a pre post (fun _ -> True) (fun _ _ _ -> True)`.

File: ulib/experimental/Steel.Channel.Simplex.fst [hunk 8]
```diff
@@ -368,7 +369,9 @@ let rec send (#p:prot) (c:chan p) (#next:prot{more next}) (x:msg_t next)
   = let v = send_receive_prelude c in //matching v as vs,vr fails
     if (fst v).chan_ctr = (snd v).chan_ctr
     then (
-      rewrite_context();
+      change_slprop (chan_inv_cond (fst v) (snd v))
+                    (pure (fst v == snd v))
+                    (fun _ -> ());
```
Comment (aseemr) [inline ulib/experimental/Steel.Channel.Simplex.fst:374]: Similar comment here, any diagnosis of requiring these changes?

File: ulib/experimental/Steel.Effect.fsti (hunk 7 above)
Comment (R1kM) [reply ulib/experimental/Steel.Effect.fsti:253]: Done

File: ulib/experimental/Steel.Channel.Simplex.fst (hunk 1 above)
Comment (R1kM) [reply ulib/experimental/Steel.Channel.Simplex.fst:164]: Done

File: ulib/experimental/Steel.Channel.Simplex.fst (hunk 2 above)
Comment (R1kM) [reply ulib/experimental/Steel.Channel.Simplex.fst:278]: This rewrite_context has always been a bit finicky, since it relies on SMT solving to prove the slprop-equivalence, it needs the shape of both slprops to match exactly.
Previously, we were "lucky" in a few cases: the inferred postcondition of the computation was matching the annotation, and no rewriting was performed (when discharging separation logic constraints, we first attempt reflexivity before jumping into the core of canon which performs AC-rewritings).

Now that we are always adding frames, even though they are often resolved to emp, shapes rarely match, which mean we have to fall back on piecewise rewritings more often

File: ulib/experimental/Steel.Effect.Atomic.fsti (hunk 3 above)
Comment (R1kM) [reply ulib/experimental/Steel.Effect.Atomic.fsti:178]: I think that if you do this, you might end up with two frames to infer for the same computation, which does not match our model or framing tactic (same problem in Steel.Effect, since the reasoning is the same).
This would happen for instance if f is already framed and you state that `if b then f else ...` is not framed. To the frame inside `f`, you would add a frame to the if_then_else when composing it with outer computations.

The better way to do this is to add additional binders/frames for f and g, and to state that the if_then_else is always framed. But this requires modifications to typecheck the if_then_else combinator w.r.t subcomp, as we discussed in the past.

File: ulib/experimental/Steel.Effect.Atomic.fsti (hunk 4 above)
Comment (R1kM) [reply ulib/experimental/Steel.Effect.Atomic.fsti:271]: Done

File: ulib/experimental/Steel.Effect.Common.fsti (hunk 5 above)
Comment (R1kM) [reply ulib/experimental/Steel.Effect.Common.fsti:1643]: This was previously needed when introducing a Steel.return. return is post-parametric, but if it has effect Steel (and not SteelF), an additional frame would have been introduced leading to two frame uvars in a constraint. The solution was this preprocessing phase, where all return post-slprops were solved before entering the core tactic, by "capturing" the full context, and solving all possible associated frames to emp.

With the current changes, return always has the framed bit set to true, which corresponds to only allowing SteelF returns in the previous formalism. Hence this preprocessing is not needed anymore

File: ulib/experimental/Steel.Effect.fsti (hunk 6 above)
Comment (R1kM) [reply ulib/experimental/Steel.Effect.fsti:225]: Cleaned up

File: ulib/experimental/Steel.Channel.Simplex.fst (hunk 8 above)
Comment (R1kM) [reply ulib/experimental/Steel.Channel.Simplex.fst:374]: Same comment as above. I didn't inspect the new constraints in detail, but I didn't expect these rewrite_context to be robust.
